### PR TITLE
Feat/view user accounts

### PR DESCRIPTION
**Pull request by AboudS777**

### CODE REVIEW

File: src/main/java/ama/account/UserController.java [hunk 1]
```diff
@@ -39,6 +43,19 @@ public String registerUserAccount(@ModelAttribute("user") @Valid User user, Bind
         return "registered";
     }
 
+    @GetMapping("/user/{username}")
+    public String viewUserAccount(@PathVariable(value = "username") String username, Model model) {
+        User user = userService.loadUserByUsername(username);
+        if (!username.equals("") && user != null) {
+            model.addAttribute("user", user);
+            model.addAttribute("submissionPosts", submissionPostRepository.findByOp(user));
+            // add attribute for liked submission posts model.addAttribute("likedPosts", submissionPostRepository.findAllByUsersWhoLiked(user));
```
Comment (SarrankanPathmanathan) [inline src/main/java/ama/account/UserController.java]: Why are we doing this?????

File: src/main/java/ama/account/UserService.java [hunk 2]
```diff
@@ -24,7 +24,7 @@ public User registerNewUserAccount(User user){
     }
 
     @Override
-    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+    public User loadUserByUsername(String username) throws UsernameNotFoundException {
```
Comment (SarrankanPathmanathan) [inline src/main/java/ama/account/UserService.java]: Needs to throw NullPointerException. I did that in my branch.

File: src/main/java/ama/post/CommentPost.java [hunk 3]
```diff
@@ -51,17 +51,13 @@ public int getDownvotes() {
     }
 
     public void upvote(User user) {
-        if(!upvoters.contains(user)) {
-            upvoters.add(user);
-            downvoters.remove(user);
-        }
+        upvoters.add(user);
```
Comment (SarrankanPathmanathan) [inline src/main/java/ama/post/CommentPost.java]: Nice!

File: src/main/java/ama/post/SubmissionPost.java [hunk 4]
```diff
@@ -2,28 +2,29 @@
...
 
     public SubmissionPost() {}
 
     public SubmissionPost(User op, String title, String text) {
         super(op, text);
         this.title = title;
-        usersWhoLiked.add(op);
     }
 
     /* title field is probably better as final*/
```
Comment (SarrankanPathmanathan) [inline src/main/java/ama/post/SubmissionPost.java]: Is it?

File: src/main/java/tests/CommentPostTests.java [hunk 5]
```diff
@@ -1,65 +1,63 @@
...
+        post.setText("Test text.");
+        submissionPostRepository.save(post);
+        post = submissionPostRepository.findAll().get(0);
         CommentPost comment = new CommentPost();
+        comment.setText("This is a test comment.");
         comment.setContext(post);
-        comment.setText("This is a test comment");
-        ResponseEntity<String> response = template.postForEntity(base.toString()+"/posts/" + URLEncoder.encode(post.toString(), "UTF-8").toString(), comment, String.class);
-        assertThat(response.getStatusCode(),equalTo(HttpStatus.FOUND));
+        mvc.perform(post("/posts/" + post.getTitle(), comment).with(user("user"))).andExpect(status().isOk());
```
Comment (AboudS777) [inline src/main/java/tests/CommentPostTests.java]: Correct me if I'm wrong but does this test a user can comment onto an AMA?

Comment (stephanernst) [inline src/main/java/tests/CommentPostTests.java]: Yes.

File: src/main/java/ama/validation/Validator.java [hunk 6]
```diff
@@ -18,6 +21,11 @@
     private SubmissionPostRepository submissionPostRepository;
 
     public void validate(User user, BindingResult result) {
+        Pattern p = Pattern.compile("[^a-z0-9]", Pattern.CASE_INSENSITIVE);
```
Comment (AboudS777) [inline src/main/java/ama/validation/Validator.java]: What does the "^" symbol for in the first parameter of Pattern.compile?

File: src/main/java/ama/post/SubmissionPost.java [hunk 7]
```diff
@@ -2,28 +2,29 @@
...
 public class SubmissionPost extends Post {
 
-    private String title;   //will need to validate title's upon form submission to ensure that there are no duplicate titles
+    private String title;
     private Date votingCloses;
     private Date answerCloses;
-    private ArrayList<User> usersWhoLiked = new ArrayList<User>();
+
+    @ManyToMany
+    private Set<User> usersWhoLiked = new HashSet<>();
```
Comment (AboudS777) [inline src/main/java/ama/post/SubmissionPost.java]: Why are you using Hashset for users who like a post?

File: src/main/java/ama/post/PostController.java [hunk 8]
```diff
@@ -41,11 +40,17 @@ public String createAMASubmission(Model model) {
     @PostMapping("/create_submission")
     public String postAMASubmission(@ModelAttribute("submissionPost") SubmissionPost post, BindingResult result){
         validator.validate(post, result);
-        if(result.hasErrors()) {
+
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        String username = auth.getName();
+        User user = userRepository.findByUsername(username);
+        if(result.hasErrors()||user == null) {
```
Comment (AboudS777) [inline src/main/java/ama/post/PostController.java]: I'm guessing this takes care of the case if a non-registered user tried to make an AMA?

File: src/main/java/tests/CommentPostTests.java (hunk 5 above)
Comment (AboudS777) [inline src/main/java/tests/CommentPostTests.java]: Correct me if I'm wrong but does this test a user can comment onto an AMA?

File: src/main/java/ama/validation/Validator.java (hunk 6 above)
Comment (AboudS777) [inline src/main/java/ama/validation/Validator.java]: What does the "^" symbol for in the first parameter of Pattern.compile?

File: src/main/java/ama/post/SubmissionPost.java (hunk 7 above)
Comment (AboudS777) [inline src/main/java/ama/post/SubmissionPost.java]: Why are you using Hashset for users who like a post?

File: src/main/java/ama/post/PostController.java (hunk 8 above)
Comment (stephanernst) [inline src/main/java/ama/post/PostController.java]: Yes.

File: src/main/java/tests/CommentPostTests.java (hunk 5 above)
Comment (stephanernst) [inline src/main/java/tests/CommentPostTests.java]: Yes.

File: src/main/java/ama/validation/Validator.java (hunk 6 above)
Comment (stephanernst) [inline src/main/java/ama/validation/Validator.java]: '^' is negation. source = https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html

File: src/main/java/ama/post/SubmissionPost.java (hunk 7 above)
Comment (stephanernst) [inline src/main/java/ama/post/SubmissionPost.java]: Users should only be able to like a post once.

File: src/main/java/ama/account/UserController.java (hunk 1 above)
Comment (RyanBillard) [inline src/main/java/ama/account/UserController.java]: stephane pls remove

File: src/main/java/ama/post/PostController.java (hunk 8 above)
Comment (RyanBillard) [inline src/main/java/ama/post/PostController.java]: but what if an ama tries to make a non-registered user????

File: src/main/java/ama/post/PostRepository.java [hunk 9]
```diff
@@ -15,5 +14,5 @@
 public interface PostRepository<T extends Post> extends CrudRepository<T, Long> {
     List<T> findByOp(User user);
     List<T> findAll();
-    List<T> findById(long id);
+    T findById(long id);
```
Comment (RyanBillard) [inline src/main/java/ama/post/PostRepository.java]: good stuff

File: src/main/java/ama/post/SubmissionPost.java (hunk 4 above)
Comment (RyanBillard) [inline src/main/java/ama/post/SubmissionPost.java]: why? And if so, why isnt it changed in this PR?

**Pull request merged**